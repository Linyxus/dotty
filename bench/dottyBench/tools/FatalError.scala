package dottyBench.tools

case class FatalError(msg: String) extends Exception(msg)
