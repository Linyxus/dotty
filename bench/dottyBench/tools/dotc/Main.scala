package dottyBench.tools
package dotc

/** Main class of the `dotc` batch compiler. */
object Main extends Driver
