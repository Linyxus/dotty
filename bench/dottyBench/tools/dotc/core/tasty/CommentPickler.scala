package dottyBench.tools.dotc.core.tasty

import dottyBench.tools.dotc.ast.tpd
import dottyBench.tools.dotc.core.Comments.{Comment, CommentsContext, ContextDocstrings}
import dottyBench.tools.dotc.core.Contexts._

import dotty.tools.tasty.TastyBuffer
import TastyBuffer.{Addr, NoAddr}

import java.nio.charset.Charset

class CommentPickler(pickler: TastyPickler, addrOfTree: tpd.Tree => Addr)(using Ctx, CState) {
  private val buf = new TastyBuffer(5000)
  pickler.newSection("Comments", buf)

  def pickleComment(root: tpd.Tree): Unit = {
    assert(ctx.docCtx.isDefined, "Trying to pickle comments, but there's no `docCtx`.")
    new Traverser(ctx.docCtx.get).traverse(root)
  }

  def pickleComment(addr: Addr, comment: Option[Comment]): Unit = comment match {
    case Some(cmt) if addr != NoAddr =>
      val bytes = cmt.raw.getBytes(Charset.forName("UTF-8"))
      val length = bytes.length
      buf.writeAddr(addr)
      buf.writeNat(length)
      buf.writeBytes(bytes, length)
      buf.writeLongInt(cmt.span.coords)
    case other =>
      ()
  }

  private class Traverser(docCtx: ContextDocstrings) extends tpd.TreeTraverser {
    override def traverse(tree: tpd.Tree)(using Ctx, CState): Unit =
      tree match {
        case md: tpd.MemberDef =>
          val comment = docCtx.docstring(md.symbol)
          pickleComment(addrOfTree(md), comment)
          traverseChildren(md)
        case _ =>
          traverseChildren(tree)
      }
  }
}
