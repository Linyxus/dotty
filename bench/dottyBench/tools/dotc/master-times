time elapsed: 18040ms
time elapsed: 6080ms
time elapsed: 4493ms
time elapsed: 4115ms
time elapsed: 4078ms
time elapsed: 3800ms
time elapsed: 3863ms
time elapsed: 3791ms
time elapsed: 3726ms
time elapsed: 3747ms
time elapsed: 3670ms
time elapsed: 3652ms
time elapsed: 3666ms
time elapsed: 3617ms
time elapsed: 3746ms
time elapsed: 3713ms
time elapsed: 3626ms
time elapsed: 3711ms
time elapsed: 3709ms
time elapsed: 3698ms
