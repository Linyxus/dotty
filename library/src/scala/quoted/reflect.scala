package scala.quoted

/** Current QuoteContext.reflect in scope
 *
 *  Usage:
 *  To use reflection inside a method
 *  ```
 *  def f(using QuoteContext) = { // or (using qctx: QuoteContext)
 *    import reflect._ // equivalent to import reflect._
 *    ...
 *  }
 *  ```
 *  or to use reflection in the signature of a method
 *  ```
 *  def f(using QuoteContext)(tree: reflect.Tree): reflect.Position = ...
 *  ```
 */
def reflect(using qctx: QuoteContext): qctx.reflect.type = qctx.reflect
