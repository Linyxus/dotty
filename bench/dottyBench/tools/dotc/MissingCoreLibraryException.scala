package dottyBench.tools.dotc

import dottyBench.tools.FatalError

class MissingCoreLibraryException(rootPackage: String) extends FatalError(
  s"""Could not find package $rootPackage from compiler core libraries.
     |Make sure the compiler core libraries are on the classpath.
   """.stripMargin
)
