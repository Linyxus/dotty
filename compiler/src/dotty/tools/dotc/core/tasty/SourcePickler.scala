package dotty.tools.dotc.core.tasty

import dotty.tools.dotc.ast.tpd
import dotty.tools.dotc.core.Comments.{Comment, CommentsContext, ContextDocstrings}
import dotty.tools.dotc.core.Contexts.Context
import dotty.tools.dotc.core.tasty.TastyBuffer.{Addr, NoAddr}

import java.nio.charset.Charset
import java.util.Base64
import java.nio.charset.StandardCharsets.UTF_8

class SourcePickler(pickler: TastyPickler)(implicit ctx: Context) {
  private val buf = new TastyBuffer(5000)
  pickler.newSection("Source", buf)

  def pickleSource(): Unit = {
    val contents = ctx.source.content.mkString
    val bytes = contents.getBytes(UTF_8)
    buf.writeInt(bytes.length)
    buf.writeBytes(bytes, bytes.length)
  }

}
