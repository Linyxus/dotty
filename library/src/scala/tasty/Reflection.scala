package scala.tasty

import scala.quoted.QuoteContext
import scala.tasty.reflect._

class Reflection(private[scala] val internal: CompilerInterface)
    extends Core
    with ConstantOps
    with ContextOps
    with CommentOps
    with FlagsOps
    with IdOps
    with ImplicitsOps
    with ImportSelectorOps
    with QuotedOps
    with PatternOps
    with PositionOps
    with Printers
    with ReportingOps
    with RootPosition
    with SignatureOps
    with StandardDefinitions
    with SymbolOps
    with TreeOps
    with TreeUtils
    with TypeOrBoundsOps { self =>

  def typeOf[T: scala.quoted.Type]: Type =
    implicitly[scala.quoted.Type[T]].unseal.tpe

  def TextEdit(pos: Position, newText: String): TextEdit =
    internal.TextEdit(pos, newText)

  def applyEdit(edit: WorkspaceEdit): Unit = internal.applyEdit(edit)

  def Result(pos: Position, label: String): Result =
    internal.Result(pos, label)

  def setResults(results: List[Result]): Unit = internal.setResults(results)

  def Completion(label: String, description: String, symbols: List[Symbol]): Completion =
    internal.Completion(label, description, symbols)

  def setCompletions(completions: List[Completion]): Unit =
    internal.setCompletions(completions)

  def setReferences(references: List[Position]): Unit =
    internal.setReferences(references)

  def Webview(title: String, body: String): Webview =
    internal.Webview(title, body)
  def setWebview(webview: Webview): Unit =
    internal.setWebview(webview)

  def allTrees(given Context): List[Tree] =
    internal.allTrees
}
