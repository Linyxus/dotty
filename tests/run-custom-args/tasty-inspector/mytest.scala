import scala.tasty.Reflection
import scala.tasty.inspector._

package tests {

class A {
  /** This is a method. */
  def method(s: String): String = s

  class AA

  object AA
}

/** Companion object to test linking */
object A

class B extends A {
  /** This is a method. */
  def otherMethod(s: String): String = s

  class BB
}

object B {
  type Z = Int
  val Z: Int = 0
}

class C 
class D[T]
class E[T] extends D[T]

class Constructors(a: String):
    def this() = this("Ala")
    def this(a: A)(b: A) = this("Ala")

/** Some methods to tests */
class Methods:
 def nobraces: A = ???
 /** Class doc test.
   * @throws [[Error]] Throws errors.
   */
 def simple(): B = ???
 def oneParam(a: A): B = ???
 def multipleParams(a: A, b: B): C = ???
 def vararg(a: A*): C = ???
 def multipleList(a: A)(b: B): C = ???

 def generic[T](a: D[T]): D[T] = ???
 def generic2[T, V](a: D[T], b: E[V]): D[T] = ???

 def primitives(a: Int, b: Double, c: Short): Byte = 0
 def strings(a: String): String = ""
 def arrays(a: Array[String], b: Array[Int]): Array[Double] = ???
}

object Test {
  def main(args: Array[String]): Unit = {
    new TestInspector().inspect("", List("tests.A"))
  }
}

class TestInspector() extends TastyInspector:

  protected def processCompilationUnit(reflect: Reflection)(root: reflect.Tree): Unit =
    import reflect._
    inspectClass(reflect)(root)

  private def inspectClass(reflect: Reflection)(tree: reflect.Tree): Unit =
    import reflect.{given _, _}
    tree match {
      case t: reflect.PackageClause =>
        def hackMembersOf(using r: Reflection)(rsym: r.Symbol) = {
          import dotty.tools.dotc
          given dotc.core.Contexts.Context = r.rootContext.asInstanceOf
          val sym = rsym.asInstanceOf[dotc.core.Symbols.Symbol]
          val members = sym.info.decls.iterator
          // println(s"members of ${sym.show} : ${members.map(_.show).mkString(", ")}")
          members.asInstanceOf[Iterator[r.Symbol]]
        }

        hackMembersOf(using reflect)(t.symbol).foreach { s =>
          println(s">>> ${s.show}")
          println(s">>> ${s.pos}")
          println(s">>> [${s.flags.show}]")
        }
        t.stats.map( m => inspectClass(reflect)(m) )
      case x =>
    }
