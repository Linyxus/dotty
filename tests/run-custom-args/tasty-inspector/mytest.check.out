>>> Gen
>>> ?
>>> [/**/]
>>> Gen
>>> compiler/test/debug/Gen.scala:<733..733>
>>> [/*final lazy object stableRealizable*/]
>>> Gen$
>>> compiler/test/debug/Gen.scala:<740..740>
>>> [/*final object moduleClass*/]
>>> Test
>>> ?
>>> [/**/]
>>> Test
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<1066..1066>
>>> [/*final lazy object stableRealizable*/]
>>> Test$
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<1073..1073>
>>> [/*final object moduleClass*/]
>>> TestInspector
>>> ?
>>> [/**/]
>>> TestInspector
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> TestInspector$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$Break
>>> ?
>>> [/**/]
>>> Gen$Break
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$Break$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$Command
>>> ?
>>> [/**/]
>>> Gen$Command
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$Command$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$Expect
>>> ?
>>> [/**/]
>>> Gen$Expect
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$Expect$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$LitExpect
>>> ?
>>> [/**/]
>>> Gen$LitExpect
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$LitExpect$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$PatExpect
>>> ?
>>> [/**/]
>>> Gen$PatExpect
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$PatExpect$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$Program
>>> ?
>>> [/**/]
>>> Gen$Program
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$Program$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$Tree
>>> ?
>>> [/**/]
>>> Gen$Tree
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$Tree$
>>> ?
>>> [/*final object moduleClass*/]
>>> Gen$EmptyExpect$
>>> ?
>>> [/**/]
>>> Gen$EmptyExpect$
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> Gen$EmptyExpect$$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.A
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<85..85>
>>> [/**/]
>>> tests.A
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<220..220>
>>> [/*final lazy object stableRealizable*/]
>>> tests.A$
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<227..227>
>>> [/*final object moduleClass*/]
>>> tests.B
>>> ?
>>> [/**/]
>>> tests.B
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<333..333>
>>> [/*final lazy object stableRealizable*/]
>>> tests.B$
>>> tests/run-custom-args/tasty-inspector/mytest.scala:<340..340>
>>> [/*final object moduleClass*/]
>>> tests.C
>>> ?
>>> [/**/]
>>> tests.C
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.C$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.Constructors
>>> ?
>>> [/**/]
>>> tests.Constructors
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.Constructors$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.D
>>> ?
>>> [/**/]
>>> tests.D
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.D$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.E
>>> ?
>>> [/**/]
>>> tests.E
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.E$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.Methods
>>> ?
>>> [/**/]
>>> tests.Methods
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.Methods$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.A$AA
>>> ?
>>> [/**/]
>>> tests.A$AA
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.A$AA$
>>> ?
>>> [/*final object moduleClass*/]
>>> tests.B$BB
>>> ?
>>> [/**/]
>>> tests.B$BB
>>> ?
>>> [/*final lazy object stableRealizable*/]
>>> tests.B$BB$
>>> ?
>>> [/*final object moduleClass*/]
